releases/releaser/releaser.go

```diff
@@ -110,44 +107,35 @@ func (r *Releaser) uploadBinaries(ctx context.Context, params PublishParameters)
 		fmt.Printf("Uploading: %q\n", key)
 
 		//#nosec:G304 // Intentionally uploading file from disk
-		in, err := os.Open(path)
+		binaryFile, err := os.Open(path)
 		if err != nil {
 			return err
 		}
-		defer closer(in, &err)
-
-		g, _ := errgroup.WithContext(ctx)
-		defer func() {
-			ferr := g.Wait()
-			if ferr != nil {
-				err = ferr
-			}
-		}()
+		defer closer(binaryFile, &err)
 
 		pr, pw := io.Pipe()
 		defer closer(pw, &err)
 
-		g.Go(func() error {
-			_, err := r.uploader.Upload(ctx, &s3.PutObjectInput{
-				Bucket:          &params.Bucket,
-				Body:            pr,
-				Key:             &key,
-				ContentEncoding: &contentEncodingGZIP,
-				ContentType:     &contentTypeOctetStream,
-
-				Tagging: encodeTags(params.Tags),
-			})
-			if err != nil {
-				_ = pw.CloseWithError(err)
-				return err
+		go func() {
```

Comment:
We use errgroup instead of go funcs to avoid leaked goroutines which error after the function has returned. Please don't use go funcs in our codebases.

Also, in this case the order of operations was carefully selected in order to correctly handle errors (specifically context cancellations), which will break with this new structure. We cannot accept this change.

Reply:
Respectfully, Errgroups do not help avoid leaked goroutines in this instance, nor is the error handling broken with this change.

Especially since in this case you do not have a group of goroutines to cancel as a group.

this change directly returns
```go
_, err = r.uploader.Upload(ctx, putObjectInput)
return err
```

If the Upload is canceled you will get the cancelation error.

Reply:
However the one about how requests are sent to S3 I think is important and should not be thrown out.

The current logic is broken. You will never get an error from S3 propagated back.

Let me explain:
- The S3 Put is in a goroutine. And it reads from the Pipereader.
- The main routine writes to the pipewriter via the io.copy to the gz.
- When the io.Copy finishes, the main function returns and the defers are run, and the pipeWriter is closed
- then the S3 request finish reading its body because the pipereader returns an EOF.

Then if the S3 error fails, you try and close the pipewriter with the S3 error, but at the point the pipe is already closed, and the main function has likely completely exited

All S3 Errors are swallowed unless they occur during the body upload

Reply:
Can you create an example of this (ideally with a test)?

Reply:
the main function can not exit until the g.Wait in the defer is released
notice that the error is returned in the  g.Go wrapped func - which will make it into the ferr of the g.Wait.
So I think we are good.
I think this can be demonstrated by adding an error...
```			err = errors.New("everything is closed - and we see this error")
```
on the original line 141 - and seeing what the tests produce